### ISSUE TITLE
[question] How do I crate a conan exe that requires a shared or static library?

### ISSUE THREAD

**Issue by RochaStratovan** [conan-io/conan#15580]
### What is your question?

 I are exploring conan on Linux.

I want to use a conan recipe to link with a static version of a library and then again with a shared version of the library.

I am creating a conan `cmake_exe` project/package for the [`zpipe`](https://www.zlib.net/zpipe.c) example. It uses the `zlib` library.

I understand I would update my simple conanfile.py recipe to add

```bash
    def requirements(self):
        self.requires("zlib/1.3.1")
```

When I run `conan install .` this fails as expected because I do not have the package installed in my local cache.

I change this to `conan install . --build=zlib/1.3.1` and now it builds the static version of the library. _**Huzzah!**_

This _completes_ half of my investigation. However, how do I instruct `conan` to build/install the `shared` version of the library?

I tried:

```bash
# clean my local cache
conan remove -c zlib

# Attempt to install/build a .so version
conan install . --build=zlib/1.3.1 -o shared=True
```

However this fails with the message:

```bash
======== Computing dependency graph ========
ERROR: option 'shared' doesn't exist
Possible options are []
```

I thought this might work based on the the zlib 1.3.1 conan-center recipe that indicates shared is an T/F option with a default of F.

Can you point me at the appropriate documentation section for this and/or teach me how to do this?

**Environment:**
* Ubuntu 18
* g++ 9.4.0
* python 3.8.10
* conan 2.0.17
* cmake 3.23.1

### Have you read the CONTRIBUTING guide?

- [X] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @RochaStratovan 

Thanks for your question.
``options`` are per-package, and they should be specified as such. Please try:
```
conan install . --build=missing -o zlib/*:shared=True
```
Recommended ``--build=missing`` in general. Note they are pattern if you want all shared dependencies you can do ``*:shared=True``. The ``-o &:shared=True`` placeholder means "the current consumer" recipe, when it applies.

**Comment by RochaStratovan** [comment]
@memsharded 

Thank you for your continued help.

I can confirm that

```bash
conan install . --build=missing -o zlib/*:shared=True
```

Created and installed the shared library dependency!!!!

I hit some additional errors that I figured out and am including in the event that others read this.

**`conan build .`** failed

```bash
ERROR: Missing binary: zlib/1.3.1:72c852c5f0ae27ca0b1741e5fd7c8b8be91a590a

zlib/1.3.1: WARN: Can't find a 'zlib/1.3.1' package binary '72c852c5f0ae27ca0b1741e5fd7c8b8be91a590a' for the configuration:
[settings]
arch=x86_64
build_type=Release
compiler=gcc
compiler.version=9
os=Linux
[options]
fPIC=True
shared=False

ERROR: Missing prebuilt package for 'zlib/1.3.1'. You can try:
    - List all available packages using 'conan list zlib/1.3.1:* -r=remote'
    - Explain missing binaries: replace 'conan install ...' with 'conan graph explain ...'
    - Try to build locally from sources using the '--build=zlib/1.3.1' argument

More Info at 'https://docs.conan.io/2/knowledge/faq.html#error-missing-prebuilt-package'
```

<br>

**`conan build . -o shared=True`** failed

```bash
======== Computing dependency graph ========
ERROR: option 'shared' doesn't exist
Possible options are []
```


<br>

**`conan build . -o zlib/*:shared=True` works**

I didn't understand that we would need to pass that along the whole way.


**TL;DR:** The following commands based on the suggested solution works:

```bash
conan install . --build=missing -o zlib/*:shared=True
conan build . -o zlib/*:shared=True
```

This is also mentioned in the documentation [Modifying options: linking the application dependencies as shared libraries](https://docs.conan.io/2/tutorial/consuming_packages/different_configurations.html#modifying-options-linking-the-application-dependencies-as-shared-libraries)

**Comment by RochaStratovan** [comment]
@memsharded,

A follow up question.

Is there a way for my executables `conanfile.py` recipe to explicitly declare that it wants a shared vs static library?

For example, suppose I had an executable `foo` and it requires

1. the static library `bar`
2. the shared library `pdq`

How would I write the `conanfile.py` recipe so that the user isn't required to pass along the build options during the install step?

**Comment by memsharded** [comment]
> conan build . -o zlib/*:shared=True

Actually, in Conan 2, all you need is:

```
conan build . --build=missing -o zlib/*:shared=True
```
It will already do the ``install`` automatically, so this one line will do it all

For the second question, yes, it is possible .In your conanfile.py you can do:
```python
class MyPkg(ConanFile):
     ...
     default_options = {"zlib/*:shared": True}
```

Note, however, that this will have less priority than command line or profiles. In Conan, the "downstream" or the user always have precedence, so if you type something like ``conan ... -o *:shared=False``, that will override the ``MyPkg`` ``default_options`` definition. Also note that defining options in many different recipes in a dependency graph can quickly get messy, because they can define different values, conflict (in reality only the first evaluation is picked). So ``options`` are better defined only in the most downstream consumer conanfiles or in ``profile`` files (or command line)

**Comment by RochaStratovan** [comment]
Thank you again for your patience and explanation.

I am confirming that I was able to follow these instructions and achieve my goal.

I appreciate the extra warnings and information on the usage too.

Cheers!

**Comment by memsharded** [comment]
Excellent, happy to help! 🙂
